File: src/tools/net-guard.js
/**
 * 网络与文件访问的**安全闸门**。
 *
 * ══════════════════════════════════════════════════════════════════
 * 为什么单独一个文件、而且写得这么啰嗦：
 *
 * 一旦 AI 能自己决定访问哪个 URL，SSRF 就是**真实**的攻击面 ——
 * 不是教科书里的。攻击方式很朴实：用户在输入框里粘一段"资料"，
 * 里面有句话诱导模型去抓 `http://169.254.169.254/latest/meta-data/`
 * （云服务的元数据接口，能拿到临时凭证），或者
 * `http://127.0.0.1:8787/api/health`（我们自己的服务）。
 *
 * 只靠提示词说"不要访问内网"是**完全不够**的。
 * 必须在真正发请求之前，用代码判断。
 *
 * 所以这里的规则是「默认全禁，逐项放开」，而不是「默认放开，禁几个」。
 * ══════════════════════════════════════════════════════════════════
 */
import net from 'node:net';
import dns from 'node:dns/promises';

export class NetGuardError extends Error {
  constructor(code, message) {
    super(message);
    this.name = 'NetGuardError';
    this.code = code;
  }
}

/** 只允许这两种协议。file:/gopher:/ftp: 之类一律拒绝。 */
const ALLOWED_PROTOCOLS = new Set(['http:', 'https:']);

/**
 * 判断一个 IP 是不是"内部地址"。
 *
 * 覆盖：
 *  · 回环 127.0.0.0/8、::1
 *  · 私有 10/8、172.16/12、192.168/16
 *  · 链路本地 169.254/16（**云元数据就在这里**）、fe80::/10
 *  · 唯一本地 fc00::/7
 *  · 运营商级 NAT 100.64/10
 *  · 未指定 0.0.0.0、::
 *  · 组播/广播
 * @param {string} ip
 * @returns {boolean}
 */
export function isPrivateIp(ip) {
  const v = String(ip ?? '').trim();
  if (!v) return true; // 判断不出来 → 当危险处理

  // IPv6
  if (net.isIPv6(v)) {
    const lower = v.toLowerCase();
    if (lower === '::1' || lower === '::') return true;
    if (lower.startsWith('fe80')) return true; // 链路本地
    if (lower.startsWith('fc') || lower.startsWith('fd')) return true; // 唯一本地 fc00::/7
    if (lower.startsWith('ff')) return true; // 组播
    // IPv4-mapped：::ffff:127.0.0.1
    //
    // ⚠️ 有两种写法，必须都认：
    //   · 点分：`::ffff:127.0.0.1`
    //   · **十六进制**：`::ffff:7f00:1`  ← Node 的 URL 会把上面那种**规范化成这种**！
    // 只认点分写法的话，`http://[::ffff:127.0.0.1]/` 就能绕过内网检查
    // （实测确认过：Node 的 URL.hostname 给的是 `[::ffff:7f00:1]`）。
    // 这是同一个 SSRF 绕过家族的第二个变体 —— 修第一个时没想到还藏着这个。
    const dotted = lower.match(/^::ffff:(\d+\.\d+\.\d+\.\d+)$/);
    if (dotted) return isPrivateIp(dotted[1]);

    const hexMapped = lower.match(/^::ffff:([0-9a-f]{1,4}):([0-9a-f]{1,4})$/);
    if (hexMapped) {
      const high = parseInt(hexMapped[1], 16);
      const low = parseInt(hexMapped[2], 16);
      const ipv4 = [high >> 8, high & 0xff, low >> 8, low & 0xff].join('.');
      return isPrivateIp(ipv4);
    }
    return false;
  }

  if (!net.isIPv4(v)) return true;

  const parts = v.split('.').map((n) => Number(n));
  if (parts.length !== 4 || parts.some((n) => !Number.isInteger(n) || n < 0 || n > 255)) {
    return true;
  }
  const [a, b] = parts;

  if (a === 0) return true; // 0.0.0.0/8
  if (a === 10) return true; // 私有
  if (a === 127) return true; // 回环
  if (a === 169 && b === 254) return true; // 链路本地 / 云元数据
  if (a === 172 && b >= 16 && b <= 31) return true; // 私有
  if (a === 192 && b === 168) return true; // 私有
  if (a === 100 && b >= 64 && b <= 127) return true; // CGNAT
  if (a === 192 && b === 0) return true; // 192.0.0.0/24 协议分配
  if (a === 198 && (b === 18 || b === 19)) return true; // 基准测试段
  if (a >= 224) return true; // 组播 + 保留 + 广播

  return false;
}

/** 这些主机名直接拒绝（连 DNS 都不用查） */
const BLOCKED_HOSTNAMES = new Set([
  'localhost',
  'localhost.localdomain',
  'ip6-localhost',
  'metadata',
  'metadata.google.internal',
  'instance-data',
]);

/**
 * 校验一个 URL 是否允许被抓取。
 *
 * @param {string} rawUrl
 * @param {object} [opts]
 * @param {string[]} [opts.allowHosts] 显式允许的主机（精确匹配或 `.example.com` 后缀）
 * @param {boolean} [opts.allowPrivateHosts] 是否允许访问内网（**默认 false**，
 *        只有用户显式打开才允许；单机场景下访问自己的服务是个正当需求，
 *        但绝不能默认开）
 * @returns {Promise<{ok:boolean, url?:URL, error?:string, resolvedIp?:string}>}
 */
export async function checkUrl(rawUrl, opts = {}) {
  const {
    allowHosts = [],
    allowPrivateHosts = false,
    // 可注入的 DNS 解析器（测试用）。
    // 为什么需要：我们的测试要断言"公网域名放行、解析到内网的域名拦下"，
    // 而这两件事都取决于**这台机器当下的 DNS**。
    // 实测踩到过：某台开发机的 DNS 会把 example.com 解析成 127.0.0.1，
    // 于是"公网地址应该放行"这条测试无缘无故就红了 —— 而代码完全没问题。
    // 靠真实 DNS 的测试是 flaky 测试，必须能注入。
    dnsLookup = (host) => dns.lookup(host, { all: true, verbatim: true }),
  } = opts;

  let url;
  try {
    url = new URL(String(rawUrl));
  } catch {
    return { ok: false, error: `这不是一个合法的网址：${String(rawUrl).slice(0, 200)}` };
  }

  if (!ALLOWED_PROTOCOLS.has(url.protocol)) {
    return {
      ok: false,
      error: `只允许 http 和 https，不允许 ${url.protocol.replace(':', '')}（这是为了安全）`,
    };
  }

  // 带凭据的 URL（http://user:pass@host）一律拒绝：这是常见的绕过手法，
  // 而且我们没有任何理由需要它。
  if (url.username || url.password) {
    return { ok: false, error: '网址里不允许带用户名和密码' };
  }

  // ⚠️ IPv6 的 hostname 在 URL 里是带方括号的：`new URL('http://[::1]/').hostname`
  // 返回的是 `"[::1]"` 而**不是** `"::1"`。
  // 不剥掉方括号的话：`net.isIP('[::1]')` 返回 0（认不出来）→ 代码会把它当**域名**
  // 去查 DNS → 查询失败或超时 → 要么放行要么报"解析不了"，
  // 总之**绕过了内网检查**。这是个真实的 SSRF 绕过（测试抓到的），
  // 而且症状是"偶发超时"而不是"被拦"，很容易被当成 flaky 忽略掉。
  const host = url.hostname
    .toLowerCase()
    .replace(/\.$/, '') // 去掉尾点（FQDN 写法）
    .replace(/^\[|\]$/g, ''); // 剥掉 IPv6 的方括号

  if (BLOCKED_HOSTNAMES.has(host)) {
    return { ok: false, error: `出于安全考虑，不允许访问内网地址（${host}）` };
  }

  const explicitlyAllowed = allowHosts.some(
    (h) => host === h.toLowerCase() || host.endsWith(`.${h.toLowerCase().replace(/^\./, '')}`),
  );

  // ⚠️ 关于 trustedHosts（对应配置里的 trustedHosts）：
  //
  // 判定"内网地址"只能靠**解析结果**，而解析结果有时会被本机环境改写。
  // 实测例子：某些开发机的 /etc/hosts 会把 github.com 指向 127.0.0.1
  // （网络管控/代理软件常这么做）。这时我们的检查会正确地拦下它 ——
  // 安全上没错，但用户会觉得"为什么 github 打不开"。
  //
  // 所以给一个**显式**的逃生口：用户可以在配置里列出他信任的域名，
  // 这些域名跳过内网 IP 检查。默认是空列表 —— 默认行为仍然是最严格的。
  // 这不是"关掉安全"，而是"把判断权交给用户，并且让他明确知道自己在做什么"。
  if (!allowPrivateHosts && !explicitlyAllowed) {
    // 主机名本身就是 IP 的情况：不用查 DNS
    if (net.isIP(host)) {
      if (isPrivateIp(host)) {
        return { ok: false, error: `出于安全考虑，不允许访问内网地址（${host}）` };
      }
      return { ok: true, url, resolvedIp: host };
    }

    // 域名：必须解析后检查真实 IP。
    // 这一步是必须的 —— 攻击者可以用一个解析到 127.0.0.1 的域名绕过纯字符串检查
    // （所谓 DNS rebinding 的简化版）。
    let addresses;
    try {
      addresses = await dnsLookup(host);
    } catch {
      return { ok: false, error: `解析不了这个域名：${host}` };
    }
    if (!addresses.length) {
      return { ok: false, error: `解析不了这个域名：${host}` };
    }
    for (const addr of addresses) {
      if (isPrivateIp(addr.address)) {
        return {
          ok: false,
          error: `出于安全考虑，不允许访问内网地址（${host} → ${addr.address}）`,
        };
      }
    }
    return { ok: true, url, resolvedIp: addresses[0].address };
  }

  return { ok: true, url };
}

/* ────────────────────────────────────────────────────────────────
 * 文件路径闸门
 * ──────────────────────────────────────────────────────────────── */

/**
 * 判断 `child` 是否在 `root` 里面（含相等）。
 * 用 path.relative 而不是字符串前缀 —— 后者会被 `root-evil` 这种绕过。
 */
export function isInsideDir(root, child, pathMod) {
  const p = pathMod;
  const rel = p.relative(p.resolve(root), p.resolve(child));
  return rel === '' || (!rel.startsWith('..') && !p.isAbsolute(rel));
}

/**
 * 校验一个路径是否允许读写。
 *
 * 规则（默认全禁）：
 *  · 必须是绝对路径（或用 allowedRoots 里的某个根拼出来的相对路径）
 *  · 解析后（含 `..` 展开、符号链接解析）必须落在某个 allowedRoot 里
 *  · 拒绝 NUL 字节
 *
 * @param {string} rawPath
 * @param {object} opts
 * @param {string[]} opts.allowedRoots 允许的根目录（绝对路径）
 * @param {import('node:fs')} opts.fs 便于测试注入
 * @param {object} opts.path node:path
 * @returns {Promise<{ok:boolean, full?:string, error?:string}>}
 */
/** 把一条路径解析成"真实路径"（展开符号链接），尽量不失败 */
async function realpathBestEffort(target, fs, p) {
  try {
    return await fs.promises.realpath(target);
  } catch {
    // 目标还不存在（比如要写新文件）：解析它的父目录，再拼回文件名
    try {
      const parent = await fs.promises.realpath(p.dirname(target));
      return p.join(parent, p.basename(target));
    } catch {
      return p.resolve(target);
    }
  }
}

/**
 * 校验一个路径是否允许读写。
 *
 * ⚠️ 关键：**允许的根目录也要 realpath**，不能只解析被检查的路径。
 *
 * 这个 bug 是被真机实测抓出来的：macOS 上 `/var` 是指向 `/private/var` 的符号链接，
 * `os.tmpdir()` 给的是 `/var/folders/...`。我们把文件解析成
 * `/private/var/.../note.txt`，却拿**未解析**的 `/var/...` 当根目录去比对 ——
 * 于是一个完全合法的文件被判定为"越权访问"。
 *
 * 更糟的是当时的错误信息把它说成"只允许访问：/var/..."，
 * 而那个目录**看起来**就是这个文件所在的目录 —— 连模型都被绕晕了
 * （它在回答里准确指出了这个矛盾）。**错误的错误信息比没有错误信息更糟。**
 *
 * 所以现在两端都解析，并且报错时把"解析后的路径"也说清楚。
 */
export async function checkPath(rawPath, { allowedRoots = [], fs, path: p } = {}) {
  const raw = String(rawPath ?? '');
  if (!raw.trim()) return { ok: false, error: '路径是空的' };
  if (raw.includes('\u0000')) return { ok: false, error: '路径里不允许有 NUL 字节' };

  if (!allowedRoots.length) {
    return {
      ok: false,
      error:
        '当前没有配置任何允许访问的目录（这是默认状态，需要你在配置文件里显式加上 allowedReadRoots）',
    };
  }

  const resolved = p.resolve(raw);
  const real = await realpathBestEffort(resolved, fs, p);

  // ⚠️ 根目录也必须解析（见上面那段 macOS 的坑）
  const realRoots = [];
  for (const root of allowedRoots) {
    realRoots.push(await realpathBestEffort(p.resolve(root), fs, p));
  }

  const inside = realRoots.some((root) => isInsideDir(root, real, p));
  if (!inside) {
    return {
      ok: false,
      error:
        `不允许访问这个路径。\n要访问的（解析后）：${real}\n允许的目录：${realRoots.join('、')}\n` +
        '（如果这个文件确实该被读到，请让用户把它放进上面某个目录，或者把该目录加进配置。）',
    };
  }
  return { ok: true, full: real };
}

File: src_py/net_guard.py
"""网络守卫: SSRF 与 DNS rebinding 防护.

这是 src/tools/net-guard.js 的 Python 移植。为什么必须有这一层:
    一旦允许工具"抓取用户给的网址", 攻击者就能让服务去访问
    169.254.169.254(云元数据, 能拿到实例凭证)或 127.0.0.1 上的内部服务。
    所以校验的不是"网址长得对不对", 而是"它最终会连到哪里"。

两个真实踩过的绕过案例(原文件注释里记着, 这里保留):
  1. **IPv4-mapped IPv6 有两种写法**: 点分 `::ffff:127.0.0.1` 与
     十六进制 `::ffff:7f00:1`。只认点分写法时,
     `http://[::ffff:127.0.0.1]/` 能绕过内网检查。
  2. **DNS rebinding**: 校验时解析成公网 IP, 连接时再解析成内网 IP。
     对策是把校验时解析到的 IP 一并返回, 让调用方直接连这个 IP。

Python 侧的一个不同: 解析后应把 host 换成已校验的 IP 再交给 urllib,
否则 urllib 会做第二次解析 —— 那就等于把 rebinding 的窗口又打开了。
"""

from __future__ import annotations

import ipaddress
import os
import socket
from pathlib import Path
from typing import Any, Callable
from urllib.parse import urlparse

ALLOWED_PROTOCOLS = {"http", "https"}

# 这些主机名直接拒绝, 连 DNS 都不用查
BLOCKED_HOSTNAMES = {
    "localhost",
    "localhost.localdomain",
    "ip6-localhost",
    "metadata",
    "metadata.google.internal",
    "instance-data",
}


class NetGuardError(Exception):
    def __init__(self, message: str, code: str = "NET_BLOCKED"):
        super().__init__(message)
        self.code = code
        self.message = message


def is_private_ip(ip: Any) -> bool:
    """判断 IP 是否属于"内部地址".

    覆盖: 回环 / 私有 / 链路本地(169.254/16 就是云元数据) / 唯一本地 /
    CGNAT / 未指定 / 组播广播。

    **判断不出来时一律当危险处理** —— 安全函数在不确定时应该保守。
    """
    v = str(ip if ip is not None else "").strip()
    if not v:
        return True
    # 去掉 IPv6 方括号
    if v.startswith("[") and v.endswith("]"):
        v = v[1:-1]
    # 去掉 zone id: fe80::1%en0
    if "%" in v:
        v = v.split("%", 1)[0]

    try:
        addr = ipaddress.ip_address(v)
    except ValueError:
        return True                      # 不是合法 IP -> 当危险

    if isinstance(addr, ipaddress.IPv6Address):
        # IPv4-mapped: 两种写法都要认
        if addr.ipv4_mapped is not None:
            return is_private_ip(str(addr.ipv4_mapped))
        # 6to4 / Teredo 里也可能藏 IPv4
        sixtofour = getattr(addr, "sixtofour", None)
        if sixtofour is not None:
            return is_private_ip(str(sixtofour))
        # IPv6: 文档段 2001:db8::/32 会被 is_private 判 True, 但它不是内网,
        # 只是保留给示例用的。放行它更符合直觉(也不会带来 SSRF 风险)。
        if addr in ipaddress.ip_network("2001:db8::/32"):
            return False
        return bool(
            addr.is_loopback or addr.is_unspecified or addr.is_link_local
            or addr.is_multicast or addr.is_private or addr.is_reserved
        )

    # 注意: Python 的 ipaddress 认为 100.64.0.0/10(CGNAT) 不是 private,
    # 而 Node 侧的实现明确拦它。这是标准库口径差异, 必须显式补上 ——
    # 否则同一份规则在两个实现里结论不同。
    cgnat = ipaddress.ip_network("100.64.0.0/10")
    if addr in cgnat:
        return True
    # 198.18.0.0/15 基准测试段同理
    if addr in ipaddress.ip_network("198.18.0.0/15"):
        return True

    return bool(
        addr.is_loopback or addr.is_unspecified or addr.is_link_local
        or addr.is_multicast or addr.is_private or addr.is_reserved
    )


def _host_matches(host: str, pattern: str) -> bool:
    """主机名匹配: 精确, 或 .example.com 后缀. 大小写不敏感."""
    h = (host or "").lower().rstrip(".")
    p = (pattern or "").lower().rstrip(".")
    if not p:
        return False
    if p.startswith("."):
        return h == p[1:] or h.endswith(p)
    return h == p


def _default_resolver(host: str, port: int) -> list[str]:
    """解析主机名 -> IP 列表. 解析失败返回空表."""
    try:
        infos = socket.getaddrinfo(host, port or None, proto=socket.IPPROTO_TCP)
    except (socket.gaierror, OSError):
        return []
    out: list[str] = []
    for info in infos:
        addr = info[4][0]
        if addr not in out:
            out.append(addr)
    return out


def check_url(raw_url: Any, *, allow_hosts: list[str] | None = None,
              allow_private_hosts: bool = False,
              resolver: Callable[[str, int], list[str]] | None = None) -> dict:
    """校验 URL 是否允许抓取.

    返回 {"ok", "url", "host", "resolved_ip", "error"}。

    allow_private_hosts 默认 False: 单机场景下访问自己的服务是正当需求,
    但绝不能默认开 —— 默认开等于默认允许访问云元数据。

    resolver 可注入(测试用): 否则"公网域名放行"这条测试会取决于
    这台机器当下的 DNS —— 实测踩到过某台开发机把 example.com 解析到 127.0.0.1,
    于是代码没问题但测试无缘无故红了。
    """
    allow_hosts = allow_hosts or []
    resolver = resolver or _default_resolver

    try:
        parsed = urlparse(str(raw_url if raw_url is not None else ""))
    except ValueError:
        return {"ok": False, "error": f"这不是一个合法的网址：{str(raw_url)[:200]}"}

    if not parsed.scheme or not parsed.netloc:
        return {"ok": False, "error": f"这不是一个合法的网址：{str(raw_url)[:200]}"}

    if parsed.scheme.lower() not in ALLOWED_PROTOCOLS:
        return {"ok": False,
                "error": f"只允许 http 和 https，不允许 {parsed.scheme}（这是为了安全）"}

    if parsed.username or parsed.password:
        return {"ok": False, "error": "网址里不允许带用户名和密码"}

    host = (parsed.hostname or "").strip()
    if not host:
        return {"ok": False, "error": "网址里没有主机名。"}

    if host.lower() in BLOCKED_HOSTNAMES:
        return {"ok": False, "error": f"出于安全考虑，不允许访问内网地址（{host}）"}

    explicitly_allowed = any(_host_matches(host, p) for p in allow_hosts)
    port = parsed.port or (443 if parsed.scheme.lower() == "https" else 80)

    if not allow_private_hosts and not explicitly_allowed:
        # 已经是 IP 的: 直接判, 不用查 DNS
        try:
            ipaddress.ip_address(host)
            if is_private_ip(host):
                return {"ok": False,
                        "error": f"出于安全考虑，不允许访问内网地址（{host}）"}
            return {"ok": True, "url": str(raw_url), "host": host,
                    "resolved_ip": host}
        except ValueError:
            pass

        addresses = resolver(host, port)
        if not addresses:
            return {"ok": False, "error": f"解析不了这个域名：{host}"}
        for addr in addresses:
            if is_private_ip(addr):
                # 只要**任一**解析结果落在内网就拒绝:
                # 攻击者可以返回多个 A 记录, 其中一个是内网。
                return {"ok": False,
                        "error": f"出于安全考虑，不允许访问内网地址（{host}）",
                        "resolved_ip": addr}
        return {"ok": True, "url": str(raw_url), "host": host,
                "resolved_ip": addresses[0]}

    return {"ok": True, "url": str(raw_url), "host": host, "resolved_ip": None}


def pinned_url(url: str, resolved_ip: str, host: str) -> str:
    """把 URL 里的主机名换成已校验的 IP, 防止第二次 DNS 解析.

    这就是防御 DNS rebinding 的关键: 校验与连接必须用**同一个** IP。
    不带 Host 头会破坏虚拟主机, 所以调用方还要补上原始 Host。
    """
    if not resolved_ip or not host:
        return url
    parsed = urlparse(url)
    netloc = parsed.netloc.replace(host, resolved_ip, 1)
    return parsed._replace(netloc=netloc).geturl()


# --------------------------------------------------------------------------
# 文件路径守卫
# --------------------------------------------------------------------------


def is_inside_dir(root: Any, child: Any) -> bool:
    """child 是否在 root 目录内(防路径穿越). 用 realpath 解析符号链接."""
    try:
        r = Path(os.path.realpath(str(root)))
        c = Path(os.path.realpath(str(child)))
    except (OSError, ValueError):
        return False
    try:
        c.relative_to(r)
        return True
    except ValueError:
        return False


def check_path(raw_path: Any, *, allowed_roots: list[str] | None = None,
               require_exists: bool = True) -> dict:
    """校验一个文件路径是否允许读取.

    默认要求路径必须落在 allowed_roots 内 —— 不能让模型随便读 ~/.ssh/id_rsa。
    """
    allowed_roots = [r for r in (allowed_roots or []) if r]
    p = str(raw_path if raw_path is not None else "").strip()
    if not p:
        return {"ok": False, "error": "没有给出文件路径。"}

    # 展开 ~ 并规范化, 否则 "~/x" 与 "/Users/x" 会被当成两个路径
    expanded = os.path.expanduser(p)
    resolved = os.path.realpath(expanded)

    if allowed_roots and not any(is_inside_dir(r, resolved) for r in allowed_roots):
        return {"ok": False,
                "error": "出于安全考虑，只能读取工作目录里的文件。"}

    if require_exists and not os.path.exists(resolved):
        return {"ok": False, "error": f"找不到这个文件：{p}"}
    if require_exists and not os.path.isfile(resolved):
        # 目录不直接报错, 但要明确说明, 免得调用方以为读到了空文件
        return {"ok": False, "error": f"这是一个目录，不是文件：{p}"}

    try:
        size = os.path.getsize(resolved) if os.path.exists(resolved) else 0
    except OSError:
        size = 0
    return {"ok": True, "path": resolved, "size": size}
